### PR TITLE
Add support for check vma in megablox kernel

### PR DESCRIPTION
**Pull request by Shuwen-Fang**
PUBLIC:
Add support for check vma in megablox kernel. This removes unnecessary collectives in the backward pass, and improves performance significantly when per device batch size is low. 

Benchmarking results (diff is with check vma enabled):
| Configuration | | TFLOP/s/dev | tokens/s/dev | TFLOP/s improvement | tokens/s improvement |
|---|---|---|---|---|---|
| High batch size | Baseline | ~260 | ~1020 | — | — |
| High batch size | Diff | ~260 | ~1030 | 0% | +1% |
| FSDP only | Baseline | ~60 | ~240 | — | — |
| FSDP only | Diff | ~80 | ~319 | +33% | +33% |
| FSDP + EP | Baseline | ~47 | ~188 | — | — |
| FSDP + EP | Diff | ~68 | ~271 | +45% | +44% |

check_vma is a jax.shard_map flag that lets callers explicitly declare which mesh axes a tensor varies over. When disabled (the default), JAX infers sharding and injects defensive all-reduces (psums) in the backward pass to guarantee correctness, even when those communications are redundant. Enabling it and passing the correct varying axes removes those unnecessary collectives.

Example command for FSDP only sharding:
```
#!/bin/bash

# --- Environment Variables ---
export CLUSTER_NAME=bodaborg-super-xpk-v54
export ZONE=us-central1-ai1a 
export PROJECT_ID=cloud-tpu-multipod-dev
export BASE_OUTPUT_DIR=gs://shuwenf-maxtext-logs 
export DEVICE_TYPE=tpu7x-256
export WORKLOAD_NAME="$(printf "%.26s" "${USER//_/-}-ds3-671b-4x4x8")-$(date +%Y%m%d-%H%M)"

# XLA Flags
XLA_FLAGS=" \
  --xla_tpu_dvfs_p_state=7 \
  --xla_tpu_scoped_vmem_limit_kib=65536 \
  --xla_tpu_bf16_emission_mode=NATIVE_EMISSION \
  --xla_tpu_enable_sparse_core_reduce_scatter_v2=true \
  --xla_tpu_enable_sparse_core_collective_offload_all_gather=true \
  --xla_tpu_enable_sparse_core_collective_offload_2d_all_gather=true \
  --xla_tpu_enable_all_gather_offload_tracing=true \
  --xla_tpu_use_tc_device_shape_on_sc=True \
  --xla_sc_disable_megacore_partitioning=True \
  --xla_tpu_enable_async_collective_fusion_fuse_all_gather=false \
  --xla_enable_async_all_gather=true \
  --xla_tpu_prefer_async_allgather_to_allreduce=true \
  --xla_tpu_enable_sparse_core_collective_offload_all_reduce=true \
  --xla_tpu_enable_sparse_core_collective_offload_reduce_scatter=true \
  --xla_tpu_enable_sparse_core_collective_offload_3d_all_gather=true \
  --xla_tpu_use_single_sparse_core_for_all_gather_offload=true \
  --xla_tpu_enable_concurrent_sparse_core_offloading=true \
  --xla_tpu_aggressive_opt_barrier_removal=true \
  --xla_tpu_enable_offloading_gather_to_sparsecore=true \
  --xla_tpu_sparse_core_all_gather_latency_multiplier=1 \
  --xla_tpu_sparse_core_reduce_scatter_latency_multiplier=3 \
  --xla_tpu_enable_sparse_core_collective_aggregator=true \
  --xla_tpu_enable_latency_hiding_layer_scheduler=true \
  --xla_tpu_scheduler_percent_shared_memory_limit=150 \
  --xla_tpu_enable_layer_scheduler_for_dependent_collectives=true \
  --xla_tpu_enable_sparse_core_collective_offload_nd_reduce_scatter=true \
  --xla_tpu_pcie_bandwidth_multiplier=0.03 \
  --xla_tpu_enable_sparse_core_offload_queuing_in_lhs=true \
  --xla_tpu_enable_multi_compute_overlap_in_layer_scheduler=false \
  --xla_tpu_enable_3d_reduce_scatter_decomposer=false "

# MaxText Workload Overrides
MAXTEXT_ARGS="\
model_name=deepseek3-671b \
per_device_batch_size=1.0 \
max_target_length=4096 \
dcn_pipeline_parallelism=1 \
dcn_data_parallelism=-1 \
ici_pipeline_parallelism=1 \
ici_fsdp_transpose_parallelism=1 \
ici_fsdp_parallelism=-1 \
allow_split_physical_axes=True \
use_iota_embed=True \
remat_policy=custom \
decoder_layer_input=offload \
opt_type=adamw \
mu_dtype=bfloat16 \
grad_dtype=bfloat16 \
use_random_routing=True \
megablox=True \
sparse_matmul=True \
use_custom_sort_vjp=True \
shard_exp_on_fsdp=True \
sa_use_fused_bwd_kernel=True \
sa_block_q=2048 \
sa_block_kv=2048 \
sa_block_q_dkv=2048 \
sa_block_kv_dkv=2048 \
sa_block_kv_dkv_compute=2048 \
sa_block_kv_dq=2048 \
sa_block_q_dq=2048 \
attention=flash \
use_tokamax_splash=True \
use_max_logit_estimate=-1 \
cost_estimate_flops_fwd=5000000000000 \
cost_estimate_flops_bwd=5000000000000 \
float32_weight_sum=False \
use_tokamax_gmm=False \
tokenizer_path=assets/tokenizer.mistral-v3 \
dataset_type=synthetic \
dataset_path=gs://max-datasets-rogue \
enable_checkpointing=False \
steps=30 \
log_config=true \
profiler=xplane \
base_output_directory=${BASE_OUTPUT_DIR} \
run_name=${WORKLOAD_NAME}"

xpk workload create \
  --cluster=$CLUSTER_NAME \
  --project=$PROJECT_ID \
  --zone=$ZONE \
  --priority=very-high \
  --max-restarts=0 \
  --device-type=tpu7x-4x4x8 \
  --num-slices=1 \
  --docker-image=gcr.io/tpu-prod-env-one-vm/chengnuojin_google_com_checkvma \
  --workload="${WORKLOAD_NAME}" \
  --command="set -e && export ENABLE_PATHWAYS_PERSISTENCE='1' && \
export LIBTPU_INIT_ARGS='${XLA_FLAGS}' && \
export TF_CPP_MIN_LOG_LEVEL=0 && \
export JAX_PLATFORMS='tpu,cpu' && export ENABLE_PJRT_COMPATIBILITY='true' && \
python3 -m MaxText.train maxtext/configs/base.yml ${MAXTEXT_ARGS}"
```

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/configs/types.py [hunk 1]
```diff
@@ -902,6 +902,11 @@ class LayoutAndSharding(BaseModel):
       le=1.0,
       description="Allowed percentage of non-sharded parameters.",
   )
+  check_vma: bool = Field(
+      False,
+      description="Enable VMA checking in shard_map calls. Recommended for improved performance but only supported "
+      "with auto sharding + EP / FSDP parallelisms.",
+  )
   shard_optimizer_over_data: bool = Field(False, description="Enable ZeRO-1 optimizer sharding over the data axis.")
```
Comment (NuojCheng) [inline src/maxtext/configs/types.py:929]: maybe throw error when check_vma is enabled but not actually used, e.g. wrong sharding combination, wrong kernel used. Also comment which cases check_vma is supported.

File: src/maxtext/kernels/megablox/backend.py [hunk 2]
```diff
@@ -522,7 +522,7 @@ def out_transform_indices(n_i, grid_id, k_i, group_metadata, group_offset):
   }
   call_gmm = qpl.pallas_call(
       kernel,
-      out_shape=jax.ShapeDtypeStruct((m, n), preferred_element_type),
+      out_shape=jax.ShapeDtypeStruct((m, n), preferred_element_type, manual_axis_type=jax.sharding.ManualAxisType()),
```
Comment (NuojCheng) [inline src/maxtext/kernels/megablox/backend.py]: I am surprised that we don't need to input physical axes names in `jax.sharding.ManualAxisType` anymore?

File: src/maxtext/layers/moe.py [hunk 3]
```diff
@@ -1150,6 +1150,17 @@ def get_quantization_dtypes():
       return lhs_quantize_dtype, rhs_quantize_dtype
 
     def gmm(inputs, kernel, tiling, group_sizes, expert_assignments, weight_gather_axes):
+      def extract_vma(tensor):
```
Comment (NuojCheng) [inline src/maxtext/layers/moe.py:1188]: could add comment explaining what this function is doing?

File: tests/integration/check_vma_test.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    "use_ring_of_experts=True",
+    "use_random_routing=false",
+    "per_device_batch_size=4",
+    "max_target_length=128",
+    "attention=flash",
+    "dtype=bfloat16",
+    "use_iota_embed=true",
+    "enable_checkpointing=false",
+    "async_checkpointing=false",
+)
```
Comment (NuojCheng) [inline tests/integration/check_vma_test.py:54]: shall we add more sharding cases? e.g. pure FSDP, pure EP, FSDP+EP(current one)

File: src/maxtext/kernels/megablox/backend.py (hunk 2 above)
Comment (Shuwen-Fang) [reply src/maxtext/kernels/megablox/backend.py]: Updated. I previously used pcast to cast axes to varying instead of ManualAxisType, but using ManualAxisType is definitely cleaner

**Review by NuojCheng — APPROVED** [review]
Great work!

File: src/maxtext/configs/types.py [hunk 5]
```diff
@@ -3199,6 +3159,23 @@ def calculate_global_batch_sizes(per_device_batch_size, expansion_factor, num_de
           "Please set use_tokamax_gmm=False."
       )
 
+    if self.check_vma:
+      if self.shard_mode != ShardMode.AUTO:
```
Comment (gobbleturk) [inline src/maxtext/configs/types.py]: style - would wrap all of this in a small function (e.g. like [here](http://google3/third_party/py/maxtext/src/maxtext/configs/types.py;l=2347;rcl=924065416)) - I realize this isn't followed in this code currently but no better time than the present to start

File: src/maxtext/configs/base.yml [hunk 6]
```diff
@@ -622,6 +609,9 @@ ici_autoregressive_parallelism: 1
 ici_pipeline_parallelism: 1
 ici_expert_parallelism: 1
 
+# Enabling check_vma is recommended for improved performance. Only supported for EP / FSDP ICI parallelisms, shard_mode: "auto", and use_tokamax_gmm=False.
```
Comment (gobbleturk) [inline src/maxtext/configs/base.yml]: are there bugs tracking tokamax support?

**Review by gobbleturk — APPROVED** [review]

File: src/maxtext/configs/base.yml (hunk 6 above)
Comment (Shuwen-Fang) [reply src/maxtext/configs/base.yml]: Created b/519282767
